src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbol.cs

```diff
@@ -251,8 +251,8 @@ internal sealed class SourcePropertySymbol : PropertySymbol, IAttributeTargetSym
 
                         // Although we only do this in error scenarios, it is undesirable to mutate the symbol by setting its type twice.
                         // Tracked by https://github.com/dotnet/roslyn/issues/35381
-                        _lazyType.InterlockedDangerousReset();
-                        _lazyType.InterlockedInitialize(type);
+                        Interlocked.Exchange(ref _lazyType, null);
```

Comment:
Should we be using a `Volatile.Write` here instead of an exchange, to ensure that we have a memory barrier?

Reply:
Talked to David to confirm that `Interlocked.Exchange` adss a memory barrier.
Looking at `pal.h` in coreclr, he said yes.
But then he also checked various mono implementations and the ARM64 JIT doesn't seem to honor that (but it's close). He'll confirm with mono folks and file an issue.

Since we don't need the old value (we're just trying to set a null), `Volatile.Write` would be fine.

However, Aleksey pointed out that we're in the constructor and so we can simplify this and just use a regular assignment.

---
In reply to: [280879267](https://github.com/dotnet/roslyn/pull/35459#discussion_r280879267) [](ancestors = 280879267)